app/controllers/credentials/credentials-form.js

```diff
@@ -0,0 +1,49 @@
+function formatErrorsForSummaryList (errors = {}) {
+  return Object.entries(errors).map(([id, message]) => ({
+    href: `#${id}`,
+    text: message
+  }))
+}
+class CredentialsForm {
+  constructor (fields = []) {
+    this.fields = fields
+    this.values = {}
+  }
+
+  from (entity = {}) {
+    const values = this.fields.reduce((aggregate, field) => {
+      const key = field.key || field.id
+      aggregate[field.id] = entity[key]
+      return aggregate
+    }, {})
+    return { values }
+  }
+
+  validate (formData = {}) {
+    const errors = {}
+    const values = {}
+
+    this.fields.forEach((field) => {
+      const valid = field.valid || []
```

Comment:
this is where typescript would be so useful! I know that it's not gonna happen anytime soon though :(